### PR TITLE
[io-layer] Extract socket

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

File: mono/io-layer/sockets.c [hunk 1]
```diff
@@ -664,55 +664,6 @@ WSAIoctl (guint32 fd, gint32 command,
 		return(SOCKET_ERROR);
 	}
 
-	if (command == SIO_KEEPALIVE_VALS) {
```
Comment (kumpera) [inline mono/io-layer/sockets.c]: How did you determine this to be dead code?

It looks like this is used by the BCL.

```
git grep 0x98000004
mcs/class/referencesource/System/net/System/Net/Sockets/IOControlCode.cs:        KeepAliveValues = 0x98000004,
mono/io-layer/sockets.h:#define SIO_KEEPALIVE_VALS 0x98000004
```

Then
```
git grep KeepAliveValues
mcs/class/System/System.Net/ServicePoint.cs:                    socket.IOControl (IOControlCode.KeepAliveValues, bytes, null);
mcs/class/referencesource/System/net/System/Net/ServicePoint.cs:                                IOControlCode.KeepAliveValues,
mcs/class/referencesource/System/net/System/Net/Sockets/IOControlCode.cs:        KeepAliveValues = 0x98000004,
```

Which suggests at least two places could be potentially using this.

File: mono/metadata/w32socket-unix.c [hunk 2]
```diff
@@ -625,6 +627,82 @@ mono_w32socket_transmit_file (SOCKET hSocket, gpointer hFile, guint32 nNumberOfB
 	return TransmitFile (hSocket, hFile, nNumberOfBytesToWrite, nNumberOfBytesPerSend, lpOverlapped, lpTransmitBuffers, dwReserved);
 }
 
+gint
+mono_w32socket_set_blocking (SOCKET socket, gboolean blocking)
```
Comment (kumpera) [inline mono/metadata/w32socket-unix.c]: I like this change. It makes explicit that we're not trying to support ioctlsocket.

**Comment by ludovic-henry** [comment]
@ntherning I am trying to get it to compile on windows, but I still get the same error ( https://jenkins.mono-project.com/job/x/2233/parsed_console/log_content.html#ERROR1 ). I don't have access to a windows box, so it's a bit hard to troubleshoot, and I have been trying to do it by updating this PR and waiting for a PR bot to build it, but it's very slow, and I didn't make much progress. So do you think you could tell me what's going wrong with these 2 errors:
```
(ClCompile target) -> 
  ..\mono\metadata\w32socket-win32.c(275): error C2059: syntax error: '{' [c:\j\workspace\w\msvc\libmonoruntime.vcxproj]
  ..\mono\metadata\w32socket-win32.c(284): error C2059: syntax error: '{' [c:\j\workspace\w\msvc\libmonoruntime.vcxproj]
```

These errors do not seem to make sense, but I suspect it's because of the way the  `WSAID_DISCONNECTEX` and `WSAID_TRANSMITFILE` are defined on windows.

**Comment by ntherning** [comment]
Here's a diff for you that will make it build:

```diff
diff --git a/mono/metadata/w32socket-win32.c b/mono/metadata/w32socket-win32.c
index a32dbd0..ef5d0d1 100644
--- a/mono/metadata/w32socket-win32.c
+++ b/mono/metadata/w32socket-win32.c
@@ -120,6 +120,11 @@ static gboolean alertable_socket_wait (SOCKET sock, int event_bit)
                blocking ? "blocking" : "non-blocking", sock, ret, _saved_error)); \
        WSASetLastError (_saved_error);

+void
+mono_w32socket_initialize (void)
+{
+}
+
 SOCKET mono_w32socket_accept (SOCKET s, struct sockaddr *addr, socklen_t *addrlen, gboolean blocking)
 {
        MonoInternalThread *curthread = mono_thread_internal_current ();
@@ -228,7 +233,6 @@ BOOL mono_w32socket_transmit_file (SOCKET hSocket, gpointer hFile, TRANSMIT_FILE
 gint
 mono_w32socket_disconnect (SOCKET sock, gboolean reuse)
 {
-       GUID disconnect_guid, transmit_file_guid;
        LPFN_DISCONNECTEX disconnect;
        LPFN_TRANSMITFILE transmit_file;
        DWORD output_bytes;
@@ -247,7 +251,7 @@ mono_w32socket_disconnect (SOCKET sock, gboolean reuse)
         * FIXME: check the best way to store pointers to functions in
         * managed objects that still works on 64bit platforms. */

-       disconnect_guid = WSAID_DISCONNECTEX;
+       GUID disconnect_guid = WSAID_DISCONNECTEX;
        ret = WSAIoctl (sock, SIO_GET_EXTENSION_FUNCTION_POINTER, &disconnect_guid, sizeof (GUID), &disconnect, sizeof (LPFN_DISCONNECTEX), &output_bytes, NULL, NULL);
        if (ret == 0) {
                if (!disconnect (sock, NULL, reuse ? TF_REUSE_SOCKET : 0, 0))
@@ -256,7 +260,7 @@ mono_w32socket_disconnect (SOCKET sock, gboolean reuse)
                return 0;
        }

-       transmit_file_guid = WSAID_TRANSMITFILE;
+       GUID transmit_file_guid = WSAID_TRANSMITFILE;
        ret = WSAIoctl (sock, SIO_GET_EXTENSION_FUNCTION_POINTER, &transmit_file_guid, sizeof (GUID), &transmit_file, sizeof (LPFN_TRANSMITFILE), &output_bytes, NULL, NULL);
        if (ret == 0) {
                if (!transmit_file (sock, NULL, 0, 0, NULL, NULL, TF_DISCONNECT | (reuse ? TF_REUSE_SOCKET : 0)))
```

**Pull request merged**